docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,196 @@
+
+There are 2 sources of data we could get –
+
+- Our dump database. We can ask the TI team to help us look at the general percent of duplicated strings on the GC heap. It would be useful to also provide tooling to help users to check how much space duplicated strings take. Some internal folks have written tooling for this but it requires work to ship.
+- We already know of teams that can benefit from deduping. We should start a conversation with them to collect some data to have a rough idea about the baseline and make sure we have a way to verify private builds with them.
+
+#### **Design outline**
+
+This is an opt in feature. When the runtime detects it’s turned on, it creates a dedup thread to do the work. 
+
+Detection of duplicated strings is done by looking into a hash table. The key into this hash table is the hash code of the content of a string. Detailed description of this detection is later in this doc. 
+
+As the dedup thread goes through the old generations linearly, it looks for references to a string object (denoted by the method table) and either calculates or looks up the hash code of that string to see if it already exists in the hash table. If so it will attempt to change the reference to point to that string with a CAS operation. If this fails, which means some other thread changed the reference at the mean time, we simply ignore this and move on. We expect the CAS failure rate to be very low.
+
+Since the new string reference we will write to the heap has the exact same type info, it means BGC can read either the old or the new reference and will get the same size info so deduping would not conflict with BGC concurrent mark. It does not conflict with BGC sweep because BGC sweep does not read references inside an object.
+
+The dedup hash table acts as weak references to the strings. Depending on the scenario we might choose to null out these weak references or not (if it’s more performant to rebuild the hash table). If we we do the former these weak references would be treated as short weak handles so the following will happen before we scan for finalization -
+
+- During BGC final mark phase we will need to null out the strings that are not marked in the hash table. This can be made concurrent. 
+
+- During a full blocking GC we will need to null out the strings that are not marked in the hash table, and relocate the ones that got promoted if we are doing a compacting GC.
+
+**Alternate design points**
+
+- Should we create multiple threads to do the work? 
+
+Deduping can be done leisurely, so it doesn’t merit having multiple threads. 
+
+- Can we use an existing thread to do the work on? 
+
+The finalizer thread is something that’s idling most of the time. However there are already plenty of types of work scheduled to potentially run on the finalizer thread so adding yet another thing, especially an opt in feature, can get messy.
```

Comment:
Also, the finalizer thread is set to run with high-priority. The string deduping is not time critical. We may want to run the string deduping thread to low priority to minimize its impact on foreground activity.

Reply:
right, I thought about mentioning but thought it was too detailed for this doc. we could change the thread priority when we do the dedup work. using a low pri thread to dedup may cause problems from the suspension perspective (I've seen scenarios where it's hard to suspend a low pri thread).